### PR TITLE
Kimi K2.5 INT4 support

### PR DESCRIPTION
**Pull request by chaitanyapk**
## Motivation

The goal of this Pull Request is to support loading and running inference on the original **Kimi 2.5 INT4 quantized weights** in both single-host and multi-host TPU environments.

Previously, there was no pipeline to unpack packed 4-bit weights dynamically on TPUs, this PR introduces an optimized loader that performs dynamic bit-unpacking on TPU and defers transpositions to post-unpacking, enabling clean compilation and execution using the Megablox GMM TPU kernels.

---

## Modifications

*   **`python/sgl_jax/srt/utils/weight_utils.py`**:
    *   Added `unpack_4bit_jax` to unpack raw `int32`/`int8` weights to `int4` on TPU.
    *   Deferred transposition of `int4` weights to TPU post-unpacking to avoid bit corruption.
*   **`python/sgl_jax/srt/layers/moe.py`**:
    *   Casts unpacked `int4` weights to `int8` right before Megablox GMM kernel compilation/execution.
    *   Fixed expert-sharded scale placeholders for static model initialization (`is_static=True`).
*   **`python/sgl_jax/test/kernels/`**:
    *   Added `test_kimi_int4_loading.py` (multi-host TP=16/EP=16) and `test_kimi_int4_loading_single.py` (TP=8/EP=1) verification tests.
    *   Added wrapper execution script `run_multihost_test.sh`.

---

## Accuracy Tests

We verified that the JAX/XLA graph compiles and executes successfully on TPU. 

*   **Test Environment**: 16 TPU devices of v7x / Ironwood (2 nodes of 4 chips each, where each chip has 2 devices).
*   **Execution Verification**: Both Dense (Layer 0 MLP) and Sparse MoE (Layer 1 EPMoE) compiled and materialized outputs successfully on TPU:
    *   **Dense FFN output shape**: `(16, 7168)`, dtype: `bfloat16`.
    *   **EPMoE GMM output shape**: `(16, 7168)`, dtype: `bfloat16`.
    *   **GMM Compiled shapes**: LHS `(128, 7168)`, RHS `(24, 7168, 2048)` (EP=16 Expert-sharded layout), scale `(24, 224, 1, 2048)`.

### Weight Loader Datatype Lifecycle Logs

Below are the trace logs verifying the correct datatype transitions from host loading to TPU compilation:

1.  **JAX TPU Bit-Unpacking**: Unpacks raw `int32` elements loaded from safetensors into JAX `int4` weights:
    ```
    INFO:root:Trace unpack_4bit_jax (before): JAX TPU array lazy_weight dtype=int32, target=int4
    INFO:root:Trace unpack_4bit_jax (after): JAX TPU array casted to final dtype=int4
    ```
2.  **Model Parameter Shape & Dtype**: After loading has finished, parameters are correctly stored in HBM with `int4` dtype:
    ```
    INFO:test_kimi_int4_loading:Parameter type after loading: <class 'jaxlib._jax.ArrayImpl'>
    INFO:test_kimi_int4_loading:Loaded Parameter shape: (384, 7168, 2048), dtype: int4
    ```
3.  **Megablox GMM Compilation Casting**: Unpacked `int4` weights are cast to `int8` container variables at compiling stage to run via GMM:
    ```
    INFO:sgl_jax.srt.layers.moe:JAX Compile Tracing GMM Stage - LHS activation shape: (128, 7168), dtype: bfloat16 | RHS weight shape: (24, 7168, 2048), dtype: int8 | RHS scale shape: (24, 224, 1, 2048), dtype: float32
    ```

For the full execution logs of the multi-host validation run, see [this GitHub Gist](https://gist.github.com/chaitanyapk/6d01a829d26e5f1404419e428b8883f6).

---

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [x] The test plan, such as providing test command.
- [ ] (Optional) The necessary documentation update.

#### 1. Single-Node Verification (TP=8, EP=1)
Run the single-host mock test:
```bash
python3 python/sgl_jax/test/kernels/test_kimi_int4_loading_single.py
```

#### 2. Distributed Multi-Host Verification (TP=16, EP=16 on 16 TPUs)
Trigger the following commands in parallel on the coordinator and worker nodes respectively:

*   **Coordinator Node (Host 0)**:
    ```bash
    rm -f /tmp/libtpu_lockfile && PYTHONPATH=python JAX_PROCESS_ID=0 JAX_COORDINATOR_ADDRESS=10.128.0.2:10025 MEGASCALE_COORDINATOR_ADDRESS=10.128.0.2:10026 python3 python/sgl_jax/test/kernels/test_kimi_int4_loading.py
    ```

*   **Worker Node (Host 1)**:
    ```bash
    rm -f /tmp/libtpu_lockfile && PYTHONPATH=python JAX_PROCESS_ID=1 JAX_COORDINATOR_ADDRESS=10.128.0.2:10025 MEGASCALE_COORDINATOR_ADDRESS=10.128.0.2:10026 python3 python/sgl_jax/test/kernels/test_kimi_int4_loading.py
    ```

### CODE REVIEW

**Comment by chaitanyapk** [comment]
@cjx0709 , @Bob-Chen222  added changes for INT4 weights for kimi k2.5 model. Please take a look and feel free to include relevant reviewers.

**Comment by cjx0709** [comment]
> @cjx0709 , @Bob-Chen222 added changes for INT4 weights for kimi k2.5 model. Please take a look and feel free to include relevant reviewers.

May we can review it, after #1331 merged.

**Comment by chaitanyapk** [comment]
Closed as a new [PR](https://github.com/sgl-project/sglang-jax/pull/1637) is tracking changes with accuracy changes done.
